File: app/src/main/java/io/github/getsixtyfour/openpyn/VpnAuthenticationHandler.java
package io.github.getsixtyfour.openpyn;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.PreferenceManager;

import com.getsixtyfour.openvpnmgmt.net.UsernamePasswordHandler;

import io.github.getsixtyfour.openpyn.security.SecurityCypher;

public final class VpnAuthenticationHandler implements UsernamePasswordHandler {

    private final Context mContext;

    public VpnAuthenticationHandler(@NonNull Context context) {
        mContext = context.getApplicationContext();
    }

    @NonNull
    public static String getHost(@NonNull Context context) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getString(context.getString(R.string.pref_openvpnmgmt_host_key),
                context.getString(R.string.pref_openvpnmgmt_host_default));
    }

    public static int getPort(@NonNull Context context) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        try {
            return Integer.parseInt(preferences.getString(context.getString(R.string.pref_openvpnmgmt_port_key),
                    context.getString(R.string.pref_openvpnmgmt_port_default)));
        } catch (NumberFormatException ignored) {
            return Integer.parseInt(context.getString(R.string.pref_openvpnmgmt_port_default));
        }
    }

    @Nullable
    public static char[] getPassword(@NonNull Context context) {
        SecurityCypher securityCypher = SecurityCypher.getInstance(context);
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return securityCypher.decrypt(preferences.getString(context.getString(R.string.pref_openvpnmgmt_password_key),
                context.getString(R.string.pref_openvpnmgmt_password_default)));
    }

    public static boolean shouldPostByteCount(@NonNull Context context) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getBoolean(context.getString(R.string.pref_openvpnmgmt_bandwidth_usage_key), true);
    }

    public static boolean shouldPostStateChange(@NonNull Context context) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getBoolean(context.getString(R.string.pref_openvpnmgmt_state_changes_key), true);
    }

    @Nullable
    private static String getUserName(@NonNull Context context) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getString(context.getString(R.string.pref_openvpnmgmt_username_key),
                context.getString(R.string.pref_openvpnmgmt_username_default));
    }

    @Nullable
    @Override
    public String getUser() {
        return getUserName(mContext);
    }

    @Nullable
    private static String getUserPass(@NonNull Context context) {
        SecurityCypher securityCypher = SecurityCypher.getInstance(context);
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return securityCypher.decryptString(preferences.getString(context.getString(R.string.pref_openvpnmgmt_userpass_key),
                context.getString(R.string.pref_openvpnmgmt_userpass_default)));
    }

    @Nullable
    @Override
    public String getPassword() {
        return getUserPass(mContext);
    }
}

File: app/src/main/kotlin/io/github/getsixtyfour/openpyn/utils/VpnAuthenticationHandler.kt
package io.github.getsixtyfour.openpyn.utils

import android.content.Context
import androidx.preference.PreferenceManager
import com.getsixtyfour.openvpnmgmt.net.UsernamePasswordHandler
import io.github.getsixtyfour.openpyn.R
import io.github.getsixtyfour.openpyn.security.SecurityCypher

class VpnAuthenticationHandler(context: Context) : UsernamePasswordHandler {

    private val mContext: Context = context.applicationContext

    override fun getUser(): String? {
        return getUserName(mContext)
    }

    override fun getPassword(): String? {
        return getUserPass(mContext)
    }

    companion object {

        fun getHost(context: Context): String {
            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
            return preferences.getString(
                context.getString(R.string.pref_openvpnmgmt_host_key), context.getString(R.string.pref_openvpnmgmt_host_default)
            )!!
        }

        fun getPort(context: Context): Int {
            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
            return runCatching {
                preferences.getString(
                    context.getString(R.string.pref_openvpnmgmt_port_key), context.getString(R.string.pref_openvpnmgmt_port_default)
                )!!
            }.fold(String::toInt) { context.getString(R.string.pref_openvpnmgmt_port_default).toInt() }
        }

        fun getPassword(context: Context): CharArray? {
            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
            val string = preferences.getString(
                context.getString(R.string.pref_openvpnmgmt_password_key), context.getString(R.string.pref_openvpnmgmt_password_default)
            )
            return if (string.isNullOrBlank()) null else SecurityCypher.getInstance(context).decrypt(string)
        }

        fun shouldPostByteCount(context: Context): Boolean {
            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
            return preferences.getBoolean(context.getString(R.string.pref_openvpnmgmt_bandwidth_usage_key), true)
        }

        fun shouldPostStateChange(context: Context): Boolean {
            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
            return preferences.getBoolean(context.getString(R.string.pref_openvpnmgmt_state_changes_key), true)
        }

        internal fun getUserName(context: Context): String? {
            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
            val string = preferences.getString(
                context.getString(R.string.pref_openvpnmgmt_username_key), context.getString(R.string.pref_openvpnmgmt_username_default)
            )
            return if (string.isNullOrBlank()) null else string
        }

        internal fun getUserPass(context: Context): String? {
            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
            val string = preferences.getString(
                context.getString(R.string.pref_openvpnmgmt_userpass_key), context.getString(R.string.pref_openvpnmgmt_userpass_default)
            )
            return if (string.isNullOrBlank()) null else SecurityCypher.getInstance(context).decryptString(string)
        }
    }
}
